mcs/bs_client.go

```diff
@@ -0,0 +1,113 @@
+package mcs
+
+import (
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/extensions/volumeactions"
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v2/snapshots"
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v2/volumes"
+)
+
+type volumeCreateOpts struct {
+	volumes.CreateOpts
+}
+
+type volumeUpdateOpts struct {
+	volumes.UpdateOpts
+}
+
+type volumeExtendSizeOpts struct {
+	volumeactions.ExtendSizeOpts
+}
+
+type volumeChangeTypeOpts struct {
+	NewType          string `json:"new_type,omitempty"`
+	AvailabilityZone string `json:"availability_zone,omitempty"`
+}
+
+type snapshotCreateOpts struct {
+	snapshots.CreateOpts
+}
+
+type snapshotUpdateMetadataOpts struct {
+	snapshots.UpdateMetadataOpts
+}
+
+type snapshotUpdateOpts struct {
+	Name        *string `json:"name,omitempty"`
+	Description *string `json:"description,omitempty"`
+}
+
+func (opts volumeChangeTypeOpts) ToVolumeChangeTypeMap() (map[string]interface{}, error) {
+	return gophercloud.BuildRequestBody(opts, "os-retype")
+}
+
+func (opts snapshotUpdateOpts) ToSnapshotUpdateMap() (map[string]interface{}, error) {
+	return gophercloud.BuildRequestBody(opts, "snapshot")
+}
+
+type snapshotUpdateResult struct {
+	commonResult
+}
+
+func (r snapshotUpdateResult) Extract() (*snapshots.Snapshot, error) {
+	var s snapshots.Snapshot
+	err := r.ExtractInto(&s)
+	return &s, err
+}
+
+func snapshotUpdateURL(c *gophercloud.ServiceClient, id string) string {
+	return c.ServiceURL("snapshots", id)
+}
+
+func volumeCreate(client *gophercloud.ServiceClient, opts volumeCreateOpts) (r volumes.CreateResult) {
+	return volumes.Create(client, opts)
+}
+
+func volumeGet(client *gophercloud.ServiceClient, id string) (r volumes.GetResult) {
+	return volumes.Get(client, id)
+}
+
+func volumeUpdate(client *gophercloud.ServiceClient, id string, opts volumeUpdateOpts) (r volumes.UpdateResult) {
+	return volumes.Update(client, id, opts)
+}
+
+func volumeDelete(client *gophercloud.ServiceClient, id string) (r volumes.DeleteResult) {
+	return volumes.Delete(client, id, nil)
+}
+
+func volumeExtendSize(client *gophercloud.ServiceClient, id string, opts volumeExtendSizeOpts) (r volumeactions.ExtendSizeResult) {
```

Comment:
How a migration policy is handled? Extend may be rejected if storage backend has not enough capacity and migration is disabled (this is an OS default)

Reply:
Are you sure that migration policy parameter is supported for volumes extensions? I haven't managed to find it in cinder code, only for retyping

Reply:
Oops, that was a scheduler hint, not an extend option.